### PR TITLE
feat(start): add api routes

### PR DESCRIPTION
**Pull request by tannerlinsley**

### CODE REVIEW

File: packages/router-generator/src/config.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ export const configSchema = z.object({
   addExtensions: z.boolean().optional().default(false),
   disableLogging: z.boolean().optional().default(false),
   disableManifestGeneration: z.boolean().optional().default(false),
+  apiBase: z.string().optional().default('/api'),
```
[the remark this answers is not in the document]
Comment (SeanCassiere) [reply packages/router-generator/src/config.ts:17]: I do kind of like the idea of splitting it up, buy I do also recognize the value in having a single place of truth for everything file-based routing.

For API routes, we need the manifest information which is nearly all the same steps we need for general fb-routing. I'd even go far as to say, we shouldn't be introducing `config` schema additions in `router-plugin` and `start` and have it all step from generator.

Maybe it's just that we need to refactor the current router-generator to be more composable. So all this logic can be split from one file into smaller ones.

**Comment by tannerlinsley** [comment]
Let’s proceed with it together for now to get it working. Then extract it out after tests are in place.

Tanner Linsley
TanStack LLC

File: packages/start/src/config/index.ts [hunk 2]
```diff
@@ -521,11 +600,164 @@ function tsrRoutesManifest(opts: {
...
+    load(id) {
+      if (id === 'tsr:api-manifest') {
+        // If we're in development, return a dummy manifest
+        // if (config.command === 'serve') {
+        //   return `export default () => ({
+        //     apiBase: '${opts.tsrConfig.apiBase}',
+        //     apiRoutes: {}
+        //   })`
+        // }
+
```
Comment (SeanCassiere) [inline packages/start/src/config/index.ts:714]: @tannerlinsley I think we need to figure out how to get this built `manifest.json` over to the `api-manifest`. Or in the `handleAPIFileRoute` function, we need to get the vinxi API manifest in there.

This is what vinxi's manifest for API looks like (located in `.vinxi/build/api/api/.vite/manifest.json`).

```json
{
  "_createApiHandler-BWUdrWLi.js": {
    "file": "assets/createApiHandler-BWUdrWLi.js",
    "name": "createApiHandler"
  },
  "_json-BfrH-og-.js": {
    "file": "assets/json-BfrH-og-.js",
    "name": "json"
  },
  "app/routes/api.boo.$id.name.$.tsx?pick=Route": {
    "file": "api.boo._id.name._.js",
    "name": "api.boo._id.name._",
    "src": "app/routes/api.boo.$id.name.$.tsx?pick=Route",
    "isEntry": true,
    "imports": [
      "_json-BfrH-og-.js",
      "_createApiHandler-BWUdrWLi.js"
    ]
  },
  "app/routes/api/foo.$.ts?pick=Route": {
    "file": "foo._.js",
    "name": "foo._",
    "src": "app/routes/api/foo.$.ts?pick=Route",
    "isEntry": true,
    "imports": [
      "_json-BfrH-og-.js",
      "_createApiHandler-BWUdrWLi.js"
    ]
  },
  "app/routes/api/hello.ts?pick=Route": {
    "file": "hello.js",
    "name": "hello",
    "src": "app/routes/api/hello.ts?pick=Route",
    "isEntry": true,
    "imports": [
      "_json-BfrH-og-.js",
      "_createApiHandler-BWUdrWLi.js"
    ]
  },
  "virtual:$vinxi/handler/api": {
    "file": "api.js",
    "name": "api",
    "src": "virtual:$vinxi/handler/api",
    "isEntry": true,
    "imports": [
      "_createApiHandler-BWUdrWLi.js"
    ]
  }
}
```

Once we get that, in the `handleAPIFileRoute` function, it'll just be a matter of handling the import (similar what's already there for server-functions).

**Comment by tannerlinsley** [comment]
You can look at the TSR:router-manifest virtual import. It brings that in.

Tanner Linsley
TanStack LLC

**Comment by SeanCassiere** [comment]
> You can look at the TSR:router-manifest virtual import. It brings that in. Tanner Linsley TanStack LLC

I was having issues with it during build, since the vinxi manifest wouldn't exist yet at. I may have a solution, but I'm still trying it out 🤞🏼.

**Comment by SeanCassiere** [comment]
🎉 API routes are working!!!

@tannerlinsley it still requires your multi-sort logic for interpolating a request URL to the relevant `routePath` (i.e `/api/user/tanner/account` -> `/api/user/$name/account`) and the parsing of the path params, but that shouldn't be too crazy since all the steps are there.

---

Notable changes.
- We don't need the "api-manifest". 
  - Using vinxi's built-in fs-router, we get around the whole manifest problem and the race-conditions that it introduced during build. 
  - We just needed to move the `createApiHandler` file into its own bundle/folder and that resolves separely.
- User-facing API imports have moved from `@tanstack/start/server` to `@tanstack/start/api`
  - This can ofc. be changed to something like `@tanstack/start/api-server` or something prior to release. Just need to make sure to update the `router-generator` template.
  - The reason it was split off is because of what was previously detailed, about needing the bundles to be separate. 
- The "api-manifest" stuff which we discussed, is no longer needed. Shall we nuke it?

**Comment by SeanCassiere** [comment]
From: https://discord.com/channels/719702312431386674/1276948324179509349

Merging in 🎉

## Changes

These are the real changes being made

### `@tanstack/router-generator`

- New option called `apiBase` that defaults to `/api`.
- Routes in the folder in `apiBase` are ignored for Start API routes.

### `@tanstack/start/config`

- New options called `routers.api.entry` that defaults to `./app/api.ts`.
- Added a new Vinxi router for handling the API routes.
- Using the `VinxiBaseFileSystemRouter` for splitting the js bundles for API routes into their own bundle.

### `@tanstack/start/api`

- New sub of the TanStack Start package for handling everything with API routes.
- Added file-based routing for API-routes.
- Added code-based routing for API routes.

## How to...

### File-based API routes

- Creating the `app/api.ts` file, with this base content for the entry handler.

```ts
// app/api.ts
import {
  createStartAPIHandler,
  defaultAPIFileRouteHandler,
} from '@tanstack/start/api'

export default createStartAPIHandler(defaultAPIFileRouteHandler)
```
- Run the dev server
- Create a file (either in dot notation or in the nested format) without the routes directory.
- Initial API route content will be automatically filled-in.

### Code-based API routes

Your files can either be split or done in a single file.

```ts
// app/api.ts
import {
  createAPIRoute,
  createStartAPIHandler,
  defaultAPIRoutesHandler,
} from '@tanstack/start/api'

const fooBarRoute = createAPIRoute('/api/foo/$bar')({
  GET: ({ params }) => {
    return new Response(`Hello ${params.bar}!`)
  },
})

export default createStartAPIHandler(
  defaultAPIRoutesHandler({
    routes: {
      '/api/foo/$bar': fooBarRoute,
      // ^ this should match the API path entered in the `fooBarRoute`
    },
  }),
)
```

If you want to split the API routes out of a single file, then this is what the `app/api.ts` file would look like.


```ts
// app/api.ts
import {
  createAPIRoute,
  createStartAPIHandler,
  defaultAPIRoutesHandler,
} from '@tanstack/start/api'
import { fooBarRoute } from './my-api-routes/foo-bar'

export default createStartAPIHandler(
  defaultAPIRoutesHandler({
    routes: { '/api/foo/$bar': fooBarRoute }
  }),
)
```

**Pull request merged**